core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -406,410 +365,312 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
+  private def getWorkspacePolicies(workspaceName: WorkspaceName): Future[Set[SamPolicyWithNameAndEmail]] = {
     for {
-      results <- resultsFuture
-      _ <- results match {
-        case RequestComplete((status: StatusCode, WorkspaceACLUpdateResponseList(usersUpdated, _, _, _))) if status.isSuccess =>
-          Future.traverse(usersUpdated) {
-            case WorkspaceACLUpdateResponse(_, email, WorkspaceAccessLevels.Write) =>
-              samDAO.addUserToPolicy(SamResourceTypeNames.billingProject, workspaceName.namespace, UserService.canComputeUserPolicyName, email, userInfo)
-            case _ => Future.successful(())
-          }
-        case RequestComplete((status: StatusCode, _)) if status.isSuccess => Future.failed(throw new RawlsException(s"unexpected message returned: $results"))
-        case _ => Future.successful(())
-      }
-    } yield results
+      workspaceId <- loadWorkspaceId(workspaceName)
+      policies <- samDAO.listPoliciesForResource(SamResourceTypeNames.workspace, workspaceId, userInfo)
+    } yield policies
   }
 
-  def sendChangeNotifications(workspaceName: WorkspaceName): Future[PerRequestMessage] = {
-
-    val getUsers = {
-      dataSource.inTransaction{ dataAccess =>
-        withWorkspaceContext(workspaceName, dataAccess) {workspaceContext =>
-          requireAccessIgnoreLock(workspaceContext.workspace, WorkspaceAccessLevels.Write, dataAccess) {
-            DBIO.sequence(workspaceContext.workspace.accessLevels.values.map {group =>
-              dataAccess.rawlsGroupQuery.flattenGroupMembership(group)
-            })
-          }
-        }
+  def validateUsersExist(userEmails: Set[String]): Future[Set[Option[String]]] = {
+    Future.traverse(userEmails) { email =>
+      samDAO.getUserIdInfo(email, userInfo).map {
+        case Left(()) => Option(email)
+        case _ => None
       }
     }
-    getUsers.map { groups =>
-      val userIds = groups.flatten.map(user => user.userSubjectId).toSet
-      val notificationMessages = userIds.map { userId => Notifications.WorkspaceChangedNotification(userId, workspaceName) }
-      val numMessages = notificationMessages.size.toString
-      notificationDAO.fireAndForgetNotifications(notificationMessages)
-      RequestComplete(StatusCodes.OK, numMessages)
-    }
-
   }
 
-  private def updateCatalogPermissions(catalogUpdates: Seq[WorkspaceCatalog], dataAccess: DataAccess, workspaceContext: SlickWorkspaceContext): ReadWriteAction[WorkspaceCatalogUpdateResponseList] = {
+  /**
+   * updates acls for a workspace
+   * @param workspaceName
+   * @param aclUpdates changes to make, if an entry already exists it will be changed to the level indicated in this
+   *                   Set, use NoAccess to remove an entry, all other preexisting accesses remain unchanged
+   * @return
+   */
+  def updateACL(workspaceName: WorkspaceName, aclUpdates: Set[WorkspaceACLUpdate], inviteUsersNotFound: Boolean): Future[PerRequestMessage] = {
```

Comment:
@MatthewBemis I took another stab at this implementation, thought you might want a look